docs/dev-guides/java-style.md

```diff
@@ -0,0 +1,166 @@
+
+### Java 8 features
+
+Use Java 8 features where helpful. See [https://leanpub.com/whatsnewinjava8/read](https://leanpub.com/whatsnewinjava8/read) and [https://www.oracle.com/java/technologies/javase/8-whats-new.html](https://www.oracle.com/java/technologies/javase/8-whats-new.html) for more details.
+
+Here is a quick run-down of Java 8 topics:
+
+*   **java.time.* – **the JavaTime library _should_ be used for date-time manipulation. Duration is especially common. For non-core extensions, consider the [https://www.threeten.org/threeten-extra/](https://www.threeten.org/threeten-extra/) library.
+*   **Lambda expressions** – _prefer_ lambda expressions for short and simple bits of code. If a lambda gets too complex, extract it into a named method.
+*   **Method references** – _prefer_ method references where possible (e.g. `someStream().map(MyClass::getBar)`).
+*   **Streams** – _prefer_ using Streams for operations that can be framed as a simple, compact stream expression. On the other hand: streams can interfere with exception handling, and complex streams may be more difficult to reason about than equivalent non-stream code, so judgement is required.
+
+### Javadoc
+
+Javadoc comments (`/** … */`) are meant to explain the intended purpose of the element they are attached to. See the [style guide](https://google.github.io/styleguide/javaguide.html) for a canonical list of when Javadoc is required.
+
+Some guidelines for writing good Javadoc:
+
+*   Spend more time documenting classes that will be frequently used or updated. Documentation mitigates against the risk of future misinterpretation.
+*   Block tags (`@param` and `@return`) are generally considered _optional_:
+    *   Use `@param` block tags when detail is required to clearly convey the meaning of each function parameter.
+    *   Use a `@return` block tag if the main text cannot clearly express the meaning of the returned value in a natural way.
+
+### Methods
+
+Some guidelines around Java methods:
+
+*   Keep methods short and simple.
+*   Avoid unwieldy parameter lists.
+*   Beware of accepting or returning null.
+*   _Prefer_ general types for parameters and specific types for return values.
+*   Avoid output parameters.
+
+See [EJ3 #49-56](https://github.com/david-sauvage/effective-java-summary#methods) for more tips.
+
+### Null
+
+Avoid nullable references where possible! Of course, this isn’t always possible, but null checks and `@Nullable` annotations should always be viewed with healthy skepticism.
+
+#### Nullable annotations
+
+@Nullable _may_ be used to indicate parameters or return values where null values are expected. When used, this annotation _should_ be in addition to (but not instead of!) Javadoc comments indicating where and when null values are expected.
+
+@NotNull and @Nonnull can be noisy; they _should not_ be used, unless there is a particular reason that readers might otherwise expect an argument to be nullable.
```

Comment:
Though, they're certainly more succinct than Javadoc saying "must not be null".

Reply:
Yep – the expectation is that "must not be null" is always implied unless explicitly stated otherwise.